File: Practicas/Hill/Hill.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

#   Criptografía y Seguridad
#   Proyecto 1: Crifrado de Hill

#   Autores:
#   Tadeo Guillén Diana G
#   Rivera González Damián


import numpy as np
import math
from Operaciones import *

alfabeto = "ABCDEFGHIJKLMNOPQRSTUVWXYZ" # Z 26

'''
    Método que dada una palabra clave y el mensaje plano a codificar
    genera el texto codificado con la matriz de la palabra clave.
    @param clave - string de palabra clave para cifrar
    @param mensaje - string texto plano para cifrar
'''
def codificar( clave, mensaje):

    rtnCadena = ""
    #1. Construcción y verificación de la Matriz correspondiente a la clave
    matrizClave = getMatrizClave(clave)

    determinante = obtenerDeterminante(matrizClave) # Obtenemos el determinante de la matriz clave
    #print("Determinante:    "+ str(determinante))
    determinanteModular = determinante%len(alfabeto)
    #print("DeterminanteModular:    "+str(determinanteModular))
    inversoDeterminante = obtenerInverso(int(determinanteModular), len(alfabeto)) # Obtenemos el inverso en Z_26    

    #2. Construcción de n-gramas del mensaje
    matrizMensaje = getMatrizMensaje(mensaje, len(matrizClave))
    #3. Multiplicamos Matrices 
    claveXmensaje = multiplicaMatrices(matrizClave, matrizMensaje, len(alfabeto))
    #4. Recuperamos mensaje ofuscado
    rtnCadena = getMessageCipher(claveXmensaje)
    return rtnCadena


'''
    Dada la palabra clave, se genera la matriz con los índices de las letras
    de la palabra, se hace la verificación que tenga inversa en Z_26
    @param clave - string
'''
def getMatrizClave(clave):

    rtnClave = []   # arreglo vacio
    clave = str.upper(clave) # pasamos a mayúsculas la clave
    raizDouble = math.sqrt( len(clave) )
    if raizDouble == int(raizDouble):

        raizInt = int(raizDouble)
        rtnClave = np.zeros((raizInt, raizInt)) # Creamos una matriz de ceros para rellenar
        flagClave = 0

        for i in range(raizInt):
            for j in range(raizInt):
                rtnClave[i][j] = alfabeto.index(clave[flagClave]) # Guardamos el índice de la letra en la matriz
                flagClave += 1


    else:   # no es exacto la raíz cuadrada del la clave
        print("Llave invalida, no se puede formar una matriz de NxN")
        quit()

    return rtnClave



'''
    NOTA: este método regresa la matriz de manera transpuesta, es decir:
    [ 2  15 13 ]
    [ 19 21 11 ]
    en lugar de :
    [ 2 19] 
    [15 21] 
    [13 11] 
    debido a que es más fácil trabajar con la primer versión de matriz
    @param mensaje - string
    @param longitudeMatrizClave - int
    @return
    '''
def getMatrizMensaje( mensaje, longitudeMatrizClave):
    try:
        rtnMensaje = []
        mensaje = completaCadena(longitudeMatrizClave, mensaje)

        if len(mensaje) % longitudeMatrizClave == 0 :
            
            rows = len(mensaje)/longitudeMatrizClave    #length of N-gramas
            rtnMensaje = np.zeros((rows, longitudeMatrizClave)) #initializing matrix

            flagPositionMesaje = 0
            
            for i in range(rows):   # i-rows
                for j in range(longitudeMatrizClave):   #j-columns                        
                    rtnMensaje[i][j] = alfabeto.index(mensaje[flagPositionMesaje])
                    flagPositionMesaje += 1
                    
        else:
            print("Mensaje invalido, no se puede formar matrices de Nx1")
        
        return rtnMensaje
    except ValueError as ve:
        print("ERROR de caracter no identificado !")
        quit()
    


'''
    Este metodo acompleta una cadena con 'X' paraque vualquier mensaje de cualquier
    tamano pueda ser codificado a cualquier clave de cualquier tamano
    @param longitudClave es la longitud del texto de la clave
    @param mensaje es el mensaje ingresado por el usuario
    @return mensaje completado para ser codificado
'''
def completaCadena( longitudClave, mensaje):
    diferencia = len(mensaje)%longitudClave
    
    if diferencia!=0:
        #print("La diferencia es: " + str(diferencia))
        for i in range(longitudClave-diferencia):
            mensaje += "X"			
    
    return mensaje;

'''
    Método que quita los caracteres que no forman parte del alfabeto
'''
def limpiaMensaje(mensaje):
    mensajeLimpio = ""
    for item in mensaje:
        if item in alfabeto:
            mensajeLimpio+=item
    
    return mensajeLimpio



'''
    Dada la matriz con los índices del mensaje cifrado o descifrado,
    genera el texto con las letras de los índices en el alfabeto.
    @param matrizChiper - matriz con índices del mensaje
'''
def getMessageCipher(matrizChiper):
    rtnMensaje = "";

    for i in range(len(matrizChiper)):
        for j in range(len(matrizChiper[i])):
            indice = int(round(matrizChiper[i][j]))%len(alfabeto)
            rtnMensaje = rtnMensaje+(alfabeto[indice]+"") 
    
    return rtnMensaje



'''
    Método que dada la matriz de los índices de las letras de la palabra clave
    y el mensaje plano a decodificar, genera el texto plano del mensaje original
    @param matrizClave - matriz de índices de las letras de la palabra clave
    @param mansaje - string mensaje cifrado para decodificar
'''
def decodificar( matrizClave, mensaje):
    rtnCadena = ""
    #1. Obtenemos la matriz inversa de la matriz clave
    matrizInversa = obtenerMatrizInversa(matrizClave, len(alfabeto))

    #2. Construcción de n-gramas del mensaje
    matrizMensaje = getMatrizMensaje(mensaje, len(matrizClave))
    #3. Multiplicamos Matrices 
    claveXmensaje = multiplicaMatrices(matrizInversa , matrizMensaje, len(alfabeto))
    #4. Recuperamos mensaje ofuscado
    rtnCadena = getMessageCipher(claveXmensaje)
    return rtnCadena


#   Autores:
#   Tadeo Guillén Diana G
#   Rivera González Damián
if __name__ == "__main__":
    

    # INGRESO DEL TEXTO A CODIFICAR
    print("\t CIFRADO DE HILL\n")
    mensaje = raw_input("Ingresa el mensaje:    ")
    mensaje = str.upper(mensaje).replace(" ", "")
    mensaje = limpiaMensaje(mensaje)

    # INGRESO DE LA PALABRA CLAVE A CODIFICAR EL TEXTO
    clave = raw_input("Ingresa la clave:    ")
    clave = str.upper(clave).replace(" ", "")
    print("\n")

    # IMPRIMIMOS LA INFORMACIÓN
    print("Mensaje original:    "+ mensaje)
    print("Clave:               "+ clave)
    print("\n")

    # Claves en Z 26
    #FZHC
    #CBFD

    # CIFRAMOS
    mensajeCodificado = codificar(clave, mensaje)
    print("\n >>>> Mensaje codificado:     "+ mensajeCodificado)
    print("\n")
    matriz_clave = getMatrizClave(clave)

    # DESCIFRAMOS
    mensajeDecodificado = decodificar(matriz_clave, mensajeCodificado)
    print("\n <<<< Mensaje decodificado:    "+ mensajeDecodificado +"\n")

File: Practicas/Hill/Hill.java
import java.math.*;
import java.util.Scanner;

public class Hill {

    static String alfabetoMayusculas = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"; 
    static int[][] mClave;
    static int[][] mClaveInversa;
    static int alfabeto;
    static int[][] matrizMensajeStatic;
    /**
     * 
     */
    private static String codificar(String clave, String mensaje) {
        String rtnCadena = "";
        //1. Construcción y verificación de la Matriz correspondiente a la clave
        int[][] matrizClave = getMatrizClave(clave);
        mClave = matrizClave; // Guardamos la matriz clave de forma estática
        alfabeto = alfabetoMayusculas.length();
        //2. Construcción de n-gramas del mensaje
        int[][] matrizMensaje = getMatrizMensaje(mensaje, matrizClave.length);
        matrizMensajeStatic = matrizMensaje;    //Guardamos las matriz del mensaje cifrado
        //3. Multiplicamos Matrices 
        int [][] claveXmensaje = multiplicaMatrices(matrizClave, matrizMensaje);

        //4. Recuperamos mensaje ofuscado
        rtnCadena = getMessageCipher(claveXmensaje);

        return rtnCadena;
    }

    /**
     * 
     * @param clave
     * @return
     */
    private static int[][] getMatrizClave(String clave){
        int[][] rtnClave = new int[0][0];

        double raizDouble = Math.sqrt( clave.length() );
        if (raizDouble == (int)raizDouble){

            int raizInt = (int)raizDouble;
            rtnClave = new int[raizInt][raizInt];
            int flagClave = 0;
            for(int i=0; i<raizInt; i++){
                for(int j=0; j<raizInt; j++){
                    rtnClave[i][j] = alfabetoMayusculas.indexOf(clave.charAt(flagClave));
                    flagClave++;
                }
            }

        }
        else{ //no es exacto la raíz cuadrada del la clave
            System.out.println("Llave invalida, no se puede formar una matriz de NxN");
            // rtnClave[i][j] = null;
        }

        return rtnClave;
    }

    /**
     * NOTA: este método regresa la matriz de manera transpuesta, es decir:
     * [ 2  15 13 ]
     * [ 19 21 11 ]
     * en lugar de :
     * [ 2 19] 
     * [15 21] 
     * [13 11] 
     * debido a que es más fácil trabajar con la primer versión de matriz y 
     * por el mismo diseño del lenguaje Java.
     * @param mensaje
     * @return
     */
    private static int[][] getMatrizMensaje(String mensaje, int longitudeMatrizClave){
        int[][] rtnMensaje = new int[0][0];
		mensaje=completaCadena(longitudeMatrizClave, mensaje);
        if( mensaje.length()%longitudeMatrizClave == 0 ){
			
            int rows = mensaje.length()/longitudeMatrizClave; //length of N-gramas
            rtnMensaje = new int[rows][longitudeMatrizClave]; //initializing matrix

            int flagPositionMesaje = 0;
            for(int i=0; i<rows; i++){ // i-rows
                for(int j=0; j<longitudeMatrizClave; j++){ // j-columns
                    rtnMensaje[i][j] = alfabetoMayusculas.indexOf(mensaje.charAt(flagPositionMesaje));
                    flagPositionMesaje++;
                    System.out.print(rtnMensaje[i][j]+" ");
                }
                System.out.print("\n");

            }

        }else{
            System.out.println("Mensaje invalido, no se puede formar matrices de Nx1");
        }

        return rtnMensaje;
    }
    
    /**
     * Este metodo acompleta una cadena con 'X' paraque vualquier mensaje de cualquier
     * tamano pueda ser codificado a cualquier clave de cualquier tamano
     * @param longitudClave es la longitud del texto de la clave
     * @param mensaje es el mensaje ingresado por el usuario
     * @return mensaje completado para ser codificado
     */
    private static String completaCadena(int longitudClave, String mensaje){
		int diferencia=mensaje.length()%longitudClave;
		if( diferencia != 0){
			System.out.println("La diferencia es:            " + diferencia);
			for(int i =0;i<(longitudClave-diferencia);i++){
				mensaje+="X";
			}
			
		}
		return mensaje;
	}

    /**
     * 
     * @param matrizClave
     * @param matrizMensaje
     * @return Matriz con operaciónes resueltas y el modulo calculado en cada posición.
     */
    private static int[][] multiplicaMatrices(int[][] matrizClave, int[][] matrizMensaje){
        int[][] rtnMultiplicacion = new int[matrizMensaje.length][matrizMensaje[0].length];
        
        for(int h=0; h<rtnMultiplicacion.length;h++){
            for(int i=0; i<matrizClave.length; i++){ // looping on rows of matrizMensaje
                int a = 0;
                for(int j=0; j<matrizMensaje[0].length; j++){
                    a = a + (matrizClave[i][j]*matrizMensaje[h][j]);
                }
                rtnMultiplicacion[h][i] = a % alfabetoMayusculas.length();
            }
        }

        return rtnMultiplicacion;
    }

    /**
     * 
     * @param matrizChiper
     * @return
     */
    private static String getMessageCipher(int[][] matrizChiper){
        String rtnMensaje = "";

        for(int i=0; i<matrizChiper.length; i++){
            for(int j=0; j<matrizChiper[i].length; j++){
                rtnMensaje = rtnMensaje + ( alfabetoMayusculas.charAt(matrizChiper[i][j]) + "");
            }
        }

        return rtnMensaje;
    }
    

    /**
     * Obtiene la matriz inversa de una matriz dentro de un Z_n
     * @param matriz - matriz de enteros
     * @param n - dimensión de Z_n
     */
    private static int[][] obtenerInversa(int[][] matriz, int n){  
        int det = Determinante.determinante(matriz);
        int[][] inversa;
        System.out.println("Tiene inversa?:"+ det%n);
        if (det%alfabeto != 1) {    // Verificamos sí la matriz tiene determinante
            System.out.println("LA MATRIZ NO TIENE INVERSA");
            return new int[0][0]; 
        }else{
            //inversa = (int[][])Inverse.invert( (double[][]) matriz );
        }
        //mClaveInversa = inversa;
        return mClaveInversa;
    }

    /**
     * 
     */
    private static String decodificar(String cadena, int[][] llave) {
        String rtnCadena = "";

        //3. Multiplicamos Matrices 
        int [][] claveXmensaje = multiplicaMatrices(mClaveInversa, matrizMensajeStatic);

        //4. Recuperamos mensaje ofuscado
        rtnCadena = getMessageCipher(claveXmensaje);

        return rtnCadena;

    }


    /**
     * Método principal
     */
    public static void main(String[] args) {

		Scanner input = new Scanner(System.in);
		System.out.println("Igresa el mensaje ");
		String mensaje = input.nextLine().toUpperCase().replaceAll(" ","");
		
		System.out.println("Igresa la clave ");
		String clave = input.nextLine().toUpperCase().replaceAll(" ","");

        // 1. Mensaje original
		System.out.println("Mensaje original: " + mensaje);
		System.out.println("Clave:            " + clave);
        System.out.println();  

        // 2. Codificar
        String mensajeCodificado = codificar(clave, mensaje);
        System.out.println("Mensaje codificado:   " + mensajeCodificado);
        //obtenerInversa(mClave, alfabeto);
        // 3. Decodificar
        //String cadenaDecodificada = decodificar(mensajeCodificado, mClaveInversa); 
        //System.out.println("Texto decodificado: " + cadenaDecodificada); 
    }

}
